apps/test/unit/storage/dataBrowser/TableControlsTest.js

```diff
@@ -0,0 +1,75 @@
+import React from 'react';
+import {Provider} from 'react-redux';
+import {mount} from 'enzyme';
+import {expect} from '../../../util/reconfiguredChai';
+
+import {
+  getStore,
+  registerReducers,
+  stubRedux,
+  restoreRedux
+} from '@cdo/apps/redux';
+import commonReducers from '@cdo/apps/redux/commonReducers';
+import {reducers} from '@cdo/apps/applab/redux/applab';
+import {changeView} from '@cdo/apps/storage/redux/data';
+import experiments from '@cdo/apps/util/experiments';
+
+import {ChartType} from '@cdo/apps/storage/dataBrowser/dataUtils';
+import TableControls from '@cdo/apps/storage/dataBrowser/TableControls';
+import VisualizerModal from '@cdo/apps/storage/dataBrowser/dataVisualizer/VisualizerModal';
+
+const DEFAULT_PROPS = {
+  clearTable: () => {},
+  exportCsv: () => {},
+  importCsv: () => {},
+  tableName: 'tableName',
+  readOnly: false
+};
+
+describe('TableControls', () => {
+  beforeEach(() => {
+    stubRedux();
+    registerReducers({...commonReducers, ...reducers});
+    experiments.setEnabled(experiments.APPLAB_DATASETS, true);
+    const store = getStore();
+    store.dispatch(changeView(DataView.TABLE), 'tableName');
+  });
+
+  afterEach(() => {
+    restoreRedux();
+    experiments.setEnabled(experiments.APPLAB_DATASETS, false);
+  });
+
+  it('creates a new VisualizerModal instance when tableName changes', () => {
+    const store = getStore();
+    const wrapper = mount(
+      <Provider store={store}>
+        <TableControls {...DEFAULT_PROPS} />
+      </Provider>
+    );
+    wrapper
+      .find(VisualizerModal)
+      .children()
+      .first()
+      .setState({chartType: ChartType.BAR_CHART, selectedColumn1: 'column'});
+    expect(
+      wrapper
+        .find(VisualizerModal)
+        .children()
+        .first()
+        .state().chartType
+    ).to.equal(ChartType.BAR_CHART);
+    wrapper.setProps({
+      children: React.cloneElement(wrapper.props().children, {
+        tableName: 'differentTable'
+      })
```

Comment:
Wow!  Looks like there were some tricky problems to solve writing this test.  Nice work.

One thing I like to explore when actions and assertions start getting verbose like this is extracting helper methods just to make the test easier to read. I've added a couple of commits to do that, and make the assertions align a little better with the idea of the state being fully reset.  Let me know what you think.